lib/src/embedded/opaque_registry.dart

```diff
@@ -0,0 +1,30 @@
+// Copyright 2019 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+/// A registry of some `T` indexed by ID so that the host can invoke
+/// them.
+final class OpaqueRegistry<T> {
+  /// Instantiations of `T` that have been sent to the host.
+  ///
+  /// The values are located at indexes in the list matching their IDs.
+  final _elementsById = <T>[];
+
+  /// A reverse map from elements to their indexes in [_elementsById].
+  final _idsByElement = <T, int>{};
+
+  /// Converts an element of type `T` to a protocol buffer to send to the host.
+  int protofy(T element) {
```

Comment:
The name `protofy` is not consistent with the other parts of the embedded compiler. AFAICT, all `protofy*` functions are returning a protocol buffer object, which is not the case here. It might be better to name it differently to avoid confusion.